include/nbl/builtin/hlsl/sort/counting.hlsl

```diff
@@ -0,0 +1,123 @@
+// Copyright (C) 2018-2024 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#include "nbl/builtin/hlsl/sort/common.hlsl"
+#include "nbl/builtin/hlsl/workgroup/arithmetic.hlsl"
+
+#ifndef _NBL_BUILTIN_HLSL_SORT_COUNTING_INCLUDED_
+#define _NBL_BUILTIN_HLSL_SORT_COUNTING_INCLUDED_
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace sort
+{
+
+NBL_CONSTEXPR uint32_t BucketsPerThread = ceil((float) BucketCount / WorkgroupSize);
+
+groupshared uint32_t prefixScratch[BucketCount];
+
+struct ScratchProxy
+{
+    uint32_t get(const uint32_t ix)
+    {
+        return prefixScratch[ix];
+    }
+    void set(const uint32_t ix, const uint32_t value)
+    {
+        prefixScratch[ix] = value;
+    }
+
+    void workgroupExecutionAndMemoryBarrier()
+    {
+        nbl::hlsl::glsl::barrier();
+    }
+};
+
+static ScratchProxy arithmeticAccessor;
+
+groupshared uint32_t sdata[BucketCount];
+
+template<typename KeyAccessor, typename ValueAccessor, typename ScratchAccessor>
+struct counting
+{
+    void histogram(NBL_REF_ARG(KeyAccessor) key, NBL_REF_ARG(ScratchAccessor) scratch, const nbl::hlsl::sort::CountingPushData data)
+    {
+        uint32_t tid = nbl::hlsl::workgroup::SubgroupContiguousIndex();
+
+        [unroll]
+        for (int i = 0; i < BucketsPerThread; i++)
+            sdata[BucketsPerThread * tid + i] = 0;
+        uint32_t index = (nbl::hlsl::glsl::gl_WorkGroupID().x * WorkgroupSize) * data.elementsPerWT;
+
+        nbl::hlsl::glsl::barrier();
```

Comment:
you should use a scratch accessor and its barrier instead of a global smem array to do the histogramming (remember you can alias the scratches)

Reply:
Done